web-api/benches/coi_personalization.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright 2023 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::{collections::HashMap, hint::black_box, time::Duration};
+
+use chrono::Utc;
+use criterion::{criterion_group, criterion_main, BatchSize, Criterion};
+use itertools::Itertools;
+use rand::{
+    distributions::{Distribution, Uniform},
+    rngs::SmallRng,
+    SeedableRng,
+};
+use xayn_ai_bert::Embedding1;
+use xayn_ai_coi::{CoiConfig, CoiId, CoiStats, PositiveCoi, UserInterests};
+use xayn_web_api::bench_rerank;
+
+macro_rules! bench_rerank {
+    ($(
+        $function: ident,
+        $embedding_size: expr,
+        $document_size: expr,
+        $interest_size: expr
+    );+ $(;)?) => {$(
+        fn $function(c: &mut Criterion) {
+            let mut rng = SmallRng::from_entropy();
+            let floats = Uniform::new_inclusive(-1.0, 1.0);
+            let ints = Uniform::new(0, $interest_size);
+
+            let system = CoiConfig::default().build();
+            let timestamp = Utc::now();
+
+            let documents = (0..$document_size)
+                .map(|_| {
+                    let embedding = Embedding1::from(
+                        floats
+                            .sample_iter(&mut rng)
+                            .take($embedding_size)
+                            .collect_vec(),
+                    )
+                    .normalize()
+                    .unwrap();
+                    let tags = vec![format!("tag {}", ints.sample(&mut rng))];
```

Comment:
:art:

maybe turn the `format!(` into a it's own function here and below to ensure we always create the same tag for the same int even if we change the code.